File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        m = len(grid)
        n = len(grid[0])
        queue = deque()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    queue.append((i, j))
        

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        while queue:
            row, col = queue.popleft()
            for dr, dc in directions:
                new_row, new_col = row +dr, col + dc
                if 0 <= new_row < m and 0 <= new_col < n and grid[new_row][new_col] == 2147483647:
                    grid[new_row][new_col] = grid[row][col] + 1
                    queue.append((new_row, new_col))

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-0.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    shortestPathBinaryMatrix(grid) {
        const n = grid.length;
        if (grid[0][0] === 1 || grid[n - 1][n - 1] === 1) {
            return -1;
        }

        const directions = [
            [0, 1], [0, -1], [1, 0], [-1, 0],
            [1, 1], [1, -1], [-1, 1], [-1, -1]
        ];

        const queue = [[0, 0, 1]];
        grid[0][0] = 1;

        while (queue.length > 0) {
            const [row, col, pathLength] = queue.shift();

            if (row === n - 1 && col === n - 1) {
                return pathLength;
            }

            for (const [dr, dc] of directions ) {
                const newRow = row + dr;
                const newCol = col + dc;

                if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < n && grid[newRow][newCol] === 0) {
                    queue.push([newRow, newCol, pathLength + 1]);
                    grid[newRow][newCol] = 1;
            }
        }
    }
    return -1;
}
}
